doc/openvpn.8

```diff
@@ -3845,6 +3846,109 @@ connection is torn down.
+switch. All client ports are in untagged mode and the tap device is
+VLAN-tagged, untagged or accepts both, depending on the
+.B \-\-vlan\-accept
+setting.
+
+Ethernet frames with a prepended 802.1Q tag are called "tagged". If the VLAN
+Identifier (VID) field in such a tag is non-zero, the frame is called
+"VLAN\-tagged". If the VID is zero, but the Priority Control Point (PCP) field
+is non\-zero, the frame is called "prio\-tagged". If there is no 802.1Q tag, the
+frame is "untagged".
+
+Using the
+.B \-\-vlan\-pvid v
+option once per client, each port can be associated with a certain VID. Packets
+can only be distributed between ports with a matching VID. Therefore, clients
+with differing VIDs are completely separated from one\-another, even if
+.B \-\-client\-to\-client
+is activated.
+
+The filtering of packets takes place in the OpenVPN server. Clients do not
+need support for VLAN tagging.
+
+The
+.B \-\-vlan\-tagging
+option is off by default. While turned off, OpenVPN
+does no parsing and accepts any Ethernet frames.
+
+The option can only be activated in
+.B \-\-dev tap
+mode.
+
+.\"*********************************************************
+.TP
+.B \-\-vlan\-accept all | tagged | untagged
+Allows the tap device's VLAN tagging policy to be configured. You can choose
+between the following modes:
+
+.B all
+(default) \-\- Admit all frames.
+.br
+.B tagged
+\-\- Admit only VLAN-tagged frames.
+.br
+.B untagged
+\-\- Admit only untagged and priority-tagged frames.
+
+(Note: Some vendors refer to switch ports running in
+.B tagged
+mode as "trunk ports" and switch ports running in
+.B untagged
+mode as "access ports".)
+
+Incoming untagged or priority\-tagged packets from clients are assigned with the
+client's Port VLAN Identifier (PVID) as their VID. In
+.B untagged
+mode, incoming untagged or priority\-tagged packets on the tap device are
+associated with the global
+.B \-\-vlan\-pvid
+setting. In
+.B tagged
+mode, any incoming untagged or priority\-tagged packets are dropped. For
+VLAN\-tagged packets, any priority information is lost as soon as the
+VLAN\-tagging is removed.
+
+In
+.B tagged
+mode, packets going out through the tap device are VLAN-tagged with the
+originating client's VID.
+
+In
+.B all
+mode, incoming tagged packets are handled the same way as in
+.B tagged
+mode. Incoming untagged packets are handled as in
+.B untagged
+mode. Outgoing packets are tagged, unless the VID matches the global PVID, in
+which case the packets go out untagged.
+.\"*********************************************************
+.TP
+.B \-\-vlan\-pvid v
+Specifies which VLAN identifier a "port" is associated with. Not valid without
+\fB\-\-vlan\-tagging\fR.
+
+In client context, the setting specifies which VLAN identifier a client is
+associated with. In global context, the tap device's VLAN identifier is set.
+The latter only makes sense in
+.B \-\-vlan\-accept untagged
+and
+.B \-\-vlan\-accept all
+mode.
+
```

Comment:
Before diving into the code I am trying to understand the whole concept behind.

My understanding is that `--vlan-pvid` can be used "tap-wide" (i.e. in the general context) and thus every received packet will be tagged with such VID (so basically all the clients are in the same VLAN), otherwise it can be used in the client context (i.e. in a CCD file) and in that case each client can have its own VID.

This is all configured on the server.
Does it mean that, when using client specific pvids, having `--vlan-accepted tagged` and a client without any `--vlan-pvid` option is simply an invalid configuration? I guess the client won't be able to send anything at all because the incoming traffic will be untagged.

Am I right?

Reply:
Hiya, I'm not the original author of these patches, @fknittel is, but I'll try my best to answer.

I believe that every client is [implicitly put onto vlan 1](https://github.com/OpenVPN/openvpn/pull/76/commits/45fb6631c316183836042852b4384445e0d0a78b#diff-78bdf3a09d213a76b88cff5b6e58e0aaR845) (the default vlan) when the server is in `--vlan-accept tagged` mode.  The client could therefore send packets, and they'd be tagged and come out of the tap adapter tagged as vlan 1.  I think that `--vlan-accept` is a server only option, and that each client can be on one and only one vlan.  Does that help explain things?

Reply:
@ordex: I'll try to give you a quick recap of the concepts (slightly reworded from the VLAN section of the patched `man -l docs/openvpn.8`):

With `--vlan-tagging` activated you might want to think of OpenVPN as a "smart switch" where the tap device is connected to one port on that switch and each client connection is connected to an additional seperate port.

In the global configuration context you basically configure the behaviour of the "tap port" while in the client context you configure the individual "client port".

`--vlan-accept` is only valid in the global context (i.e. for the tap port), (all client ports are hard-coded to `--vlan-accept untagged`).

`--vlan-accept tagged` means that the port only accepts tagged packets. All untagged packets are dropped.

`--vlan-pvid` is only relevant in the case where the port (also) accepts untagged packets. (E.g. on client ports or on the tap port with `--vlan-accept all` or `--vlan-accept untagged`.) Read PVID as "port's virtual lan id that is applied to all packets that don't have a virtual lan id". All untagged packets get tagged with the PVID. All accepted tagged packets keep their virtual lan id.

ikelos is correct: If a client has no  `--vlan-pvid` specified, it defaults to 1 and a client can only participate in a single VLAN. This has nothing to do with the `--vlan-accept` mode of the tap port though. In `--vlan-tagging` mode, all clients are expected to send/receive untagged packets, so they can't possibly support multiple VLANs at the same time.

HTH?

Reply:
Hi @fknittel , thanks for taking some time to explain this further.

> In the global configuration context you basically configure the behaviour of the "tap port" while in the client context you configure the individual "client port".

Oh ok, this clarifies a lot. The meaning of the global context was not clear, but now it is. Because basically the tap interface is like yet another L2 client, so it gets its own settings.

Ok, the rest makes sense. Thanks!